### PR TITLE
[DirectX] Get resource information via TargetExtType

### PR DESCRIPTION
**Pull request by bogner**
Instead of storing an auxilliary structure with the information from the DXIL resource target extension types duplicated, access the information that we can via the type itself.

This also means we need to handle some of the target extension types we haven't fully defined yet, like Texture and CBuffer. For now we make an educated guess to what those should look like based on llvm/wg-hlsl#76, and we can update them fairly easily when we've defined them more thoroughly.

First part of #118400

### CODE REVIEW

**Comment by bogner** [comment]
> ⚠️ undef deprecator found issues in your code. ⚠️
> 
>  You can test this locally with the following command:
> The following files introduce new uses of undef:
> 
> * llvm/unittests/Analysis/DXILResourceTest.cpp

These are all resolved in #119775, and really this is just exposing that we were already doing the wrong thing here. I don't think there's anything that needs to be done in this PR.

File: llvm/include/llvm/Analysis/DXILResource.h [hunk 1]
```diff
@@ -11,19 +11,191 @@
...
+///
+/// `target("dx.CBuffer", <Type>, ...)`
+class CBufferExtType : public TargetExtType {
+public:
+  CBufferExtType() = delete;
+  CBufferExtType(const CBufferExtType &) = delete;
+  CBufferExtType &operator=(const CBufferExtType &) = delete;
+
+  Type *getResourceType() const { return getTypeParameter(0); }
+
```
Comment (hekota) [inline llvm/include/llvm/Analysis/DXILResource.h:167]: First argument of the CBuffer target type is most likely going to be its size:
```
bool getCBufferSize() const { return getIntParameter(0); }
```

File: llvm/lib/Analysis/DXILResource.cpp [hunk 2]
```diff
@@ -197,184 +258,128 @@ bool ResourceInfo::isMultiSample() const {
...
-  if (isTyped() && RHS.isTyped() && Typed < RHS.Typed)
-    return true;
-  if (isMultiSample() && RHS.isMultiSample() && MultiSample < RHS.MultiSample)
-    return true;
-  return false;
+uint32_t ResourceInfo::getCBufferSize(const DataLayout &DL) const {
+  assert(isCBuffer() && "Not a CBuffer");
+  Type *Ty = cast<CBufferExtType>(HandleTy)->getResourceType();
+  return DL.getTypeSizeInBits(Ty) / 8;
+}
```
Comment (hekota) [inline llvm/lib/Analysis/DXILResource.cpp:301]:
```suggestion
uint32_t ResourceInfo::getCBufferSize() const {
  assert(isCBuffer() && "Not a CBuffer");
  return cast<CBufferExtType>(HandleTy)->getCBufferSize();
}
```

**Review by hekota — APPROVED** [review]
LGTM! Lots of code churn, but fairly easy to follow. One small suggestion related to cbuffers which can be addressed later.

File: llvm/include/llvm/Analysis/DXILResource.h [hunk 3]
```diff
@@ -11,19 +11,192 @@
...
 #include "llvm/Pass.h"
 #include "llvm/Support/Alignment.h"
 #include "llvm/Support/DXILABI.h"
 
 namespace llvm {
 class CallInst;
+class DataLayout;
 class LLVMContext;
 class MDTuple;
+class TargetExtType;
```
Comment (llvm-beanz) [inline llvm/include/llvm/Analysis/DXILResource.h:25]: You can't derive from a forward-declared type so you must be including the header or depending on an include ordering.

**Review by llvm-beanz — APPROVED** [review]
One nit, otherwise LGTM.

File: llvm/include/llvm/Analysis/DXILResource.h (hunk 3 above)
Comment (bogner) [reply llvm/include/llvm/Analysis/DXILResource.h]: Good point. I've replaced this with an explicit include of DerivedTypes.h

**Pull request merged**